### PR TITLE
Add advanced TLS settings to harden setup.

### PR DESCRIPTION
**Pull request by roman-vynar**
Added advanced TLS settings to `[server]` section to harden Grafana setup.

Currently, there is no possibility to restrict Grafana to the specific TLS version, e.g. `1.2` and the ciphersuites to use. Thus, it makes the setup vulnerable to possible SSL/TLS attacks, for example SWEET32 attack (see below). 

Golang defines 17 ciphersuites, 12 of them have grade A, the other 5 (RC4, 3DES) - grade C. By default, RC4 are excluded but 3DES not. Adding the additional config options can help to exclude those ciphers and harden your setup when Grafana is exposed to the public internet.

This is important as Grafana contains datasource credentials even not exposed explicitly, not counting the actual data whatever they are.

Before:
```
$ nmap --script +ssl-enum-ciphers -p 3000 127.0.0.1

Starting Nmap 7.40 ( https://nmap.org ) at 2017-01-24 15:00 EET
Nmap scan report for localhost (127.0.0.1)
Host is up (0.00024s latency).
PORT     STATE SERVICE
3000/tcp open  ppp
| ssl-enum-ciphers: 
|   TLSv1.0: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (secp256r1) - C
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|     compressors: 
|       NULL
|     cipher preference: client
|     warnings: 
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|   TLSv1.1: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (secp256r1) - C
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|     compressors: 
|       NULL
|     cipher preference: client
|     warnings: 
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (secp256r1) - C
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (secp256r1) - A
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|     compressors: 
|       NULL
|     cipher preference: client
|     warnings: 
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|_  least strength: C

Nmap done: 1 IP address (1 host up) scanned in 0.43 seconds
```

Once you configure the following
```
[server]
...
tls_min_version = tls12
tls_cipher_suites = TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA
tls_prefer_server_cipher_suites = true
```
you will get a high grade of security:
```
$ nmap --script +ssl-enum-ciphers -p 3000 127.0.0.1

Starting Nmap 7.40 ( https://nmap.org ) at 2017-01-24 15:01 EET
Nmap scan report for localhost (127.0.0.1)
Host is up (0.00022s latency).
PORT     STATE SERVICE
3000/tcp open  ppp
| ssl-enum-ciphers: 
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|     compressors: 
|       NULL
|     cipher preference: server
|_  least strength: A

Nmap done: 1 IP address (1 host up) scanned in 0.41 seconds
```

### CODE REVIEW

**Comment by bergquist** [comment]
ref https://github.com/grafana/grafana/pull/6838

**Comment by roman-vynar** [comment]
@bergquist Hi Carl, https://github.com/grafana/grafana/pull/6838 is really a different topic.

File: conf/defaults.ini [hunk 1]
```diff
@@ -57,6 +57,17 @@ enable_gzip = false
 cert_file =
 cert_key =
 
+# TLS settings.
+# The minimum supported version of TLS. Either "tls10", "tls11" or "tls12", default is "tls10".
+tls_min_version = tls10
```
Comment (fg2it) [inline conf/defaults.ini]: at the very least, default min version should be 1.1 and probably 1.2 would be better.

Comment (rhuddleston) [inline conf/defaults.ini]: Prior to this change tls10 was the default, after it's merged you can change it to tls12 as the default, which I also recommend, but may not work for everyone.

**Comment by torkelo** [comment]
I think these options are best handled by a reverse proxy like nginx or apache. It is the preferred / recommended way to expose a grafana-server to the interenet. Nginx or Apache have thousands of features and patches aimed at security. 

Can't see options like this for MongoDB or Elasticsearch. But maybe we can improve the default? Would changing the TLS min version break for many users?

**Comment by roman-vynar** [comment]
Yes, the best way is to put security behind nginx/apache. However, if Grafana offers TLS out of the box then it would be great to have those basic options to harden the setup.

I think TLS 1.2 should be set be default. Due to Grafana is mostly used from the web browsers, it should be ok. Who has TLS configured and for some reasons TLS 1.2 does not work for them, they can change the setting.

**Comment by torkelo** [comment]
but I would prefer not to introduce a setting and bloat config files and config docs.

**Comment by roman-vynar** [comment]
We can remove config/docs options and leave the best defaults, TLS 1.2, all grade A ciphers.

**Comment by bergquist** [comment]
I like the idea of making secure options the default. Seems like most browsers support tls 1.2 http://caniuse.com/#feat=tls1-2 

Those we cannot use tls 1.2 for some reason could use nginx infront of grafana.

**Comment by torkelo** [comment]
yea, think having a hard coded TLS 1.2 as min version makes sense. If if people want less restrictive SSL they can disable SSL in grafana and use nginx

**Comment by roman-vynar** [comment]
Okay, hardcoded the most secured TLS options available and removed config/docs options:
```
weber@pluto:~$ nmap --script +ssl-enum-ciphers -p 3000 127.0.0.1

Starting Nmap 7.40 ( https://nmap.org ) at 2017-02-02 13:39 EET
Nmap scan report for localhost (127.0.0.1)
Host is up (0.00032s latency).
PORT     STATE SERVICE
3000/tcp open  ppp
| ssl-enum-ciphers: 
|   TLSv1.2: 
|     ciphers: 
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (secp256r1) - A
|     compressors: 
|       NULL
|     cipher preference: server
|_  least strength: A

Nmap done: 1 IP address (1 host up) scanned in 0.37 seconds
```

**Comment by bergquist** [comment]
Thanks for contributing! :)

**Comment by tommyorndorff** [comment]
Can someone tell me what I'm doing wrong here?  No matter what I try, I'm offered TLSv1.0 up to TLSv1.2.  With PCI 3 requirements coming, my enterprise is requiring TLSv1.2 only.

I'm running a stock Centos7 ec2 instance from the AWS Marketplace, if that matters.  

```
[centos@ip-10-15-22-240 ~]$ grafana-server -v
Version 4.3.2 (commit: ed4d170)
[centos@ip-10-15-22-240 ~]$ nmap --script +ssl-enum-ciphers -p 3000 127.0.0.1

Starting Nmap 6.40 ( http://nmap.org ) at 2017-06-15 00:37 UTC
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000043s latency).
PORT     STATE SERVICE
3000/tcp open  ppp
| ssl-enum-ciphers:
|   TLSv1.0:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|     compressors:
|       NULL
|   TLSv1.1:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|     compressors:
|       NULL
|   TLSv1.2:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_GCM_SHA256 - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_GCM_SHA384 - strong
|     compressors:
|       NULL
|_  least strength: strong

Nmap done: 1 IP address (1 host up) scanned in 0.06 seconds
```

**Comment by daniellee** [comment]
@tommyorndorff I seem to be seeing the same problem. Investigating it now.

**Comment by tommyorndorff** [comment]
Thank you for the second confirmation.  I've opened issue #8641 for tracking.

**Comment by daniellee** [comment]
@tommyorndorff confirmed this as a bug. A subsequent change to http_server.go caused the code that sets the TLS config to not be called.

**Pull request merged**